File: src/Flag.cpp
#include <cstdio>
#include <string>
#include "common_flags.h"

class Flag {
private:
    int buffer[8];
    FILE *flag_f;

public:
    Flag(std::string a, std::string b) {
        flag_f = fopen(a.c_str(), b.c_str());
    }

    ~Flag() {
        fclose(flag_f);
    }

    void refreash(std::string a, std::string b) {
        flag_f = fopen(a.c_str(), b.c_str());
    }
    
    void del() {
        fclose(flag_f);
    }

    int *get_data() {
        _rewind();
        for (size_t i = 0; i < 8; i++)
            fscanf(flag_f, "%d", &buffer[i]);
        return buffer;
    }

    void set_stat(int b[8]) {
        _rewind();
        for (size_t i = 0; i < 8; i++)
            fprintf(flag_f, "%d ", b[i]);
    }

    void set_stat(int self, int stat, int id) {
        int b[] = {self, stat, id, 0, 0, 0, 0, 0};
        set_stat(b);
    }

    void set_stat(int self, int stat) {
        int b[] = {self, stat, 0, 0, 0, 0, 0, 0};
        set_stat(b);
    }

    int get_requestor_id() {
        return buffer[0];
    }

    bool is_requested() {
        return buffer[1] == REQUESTED;
    }

    bool is_answered() {
        return buffer[1] == ANSWERED;
    }

    bool is_acquired() {
        return buffer[1] == ACQUIRED;
    }

    void _rewind() {
        rewind(flag_f);
    }
};

File: src_achv/Flag.py
from common_flags import Stat


class Flag():

    def __init__(self, f):
        self.flag_f = f

    def __del__(self):
        self.flag_f.close()

    def get_data(self):
        self.rewind()
        self.buffer = self.flag_f.readline().split()
        return self.buffer

    def set_stat(self, data):
        self.rewind()
        for i in range(0, 8):
            self.flag_f.write("{0} ".format(data[i], end=""))

    def get_requestor_id(self):
        if not self.buffer:
            return False
        return self.buffer[0]

    def get_target_id(self):
        if not self.buffer:
            return False
        return self.buffer[2]

    def is_acquired(self):
        if not self.buffer:
            return False
        return int(self.buffer[1]) == Stat.ACQUIRED

    def is_requested(self):
        if not self.buffer:
            return False
        return int(self.buffer[1]) == Stat.REQUESTED

    def is_answered(self):
        if not self.buffer:
            return False
        return int(self.buffer[1]) == Stat.ANSWERED

    def is_busy(self):
        if not self.buffer:
            return False
        return int(self.buffer[1]) == Stat.BUSY

    def rewind(self):
        self.flag_f.seek(0)
